crates/bindings-macro/src/lib.rs

```diff
@@ -626,119 +626,73 @@ fn spacetimedb_tabletype_impl(item: syn::DeriveInput) -> syn::Result<TokenStream
-                Ok(col.index)
+                Ok((col.index, col.field.ty))
             })
+            // TODO: stabilized in 1.79
+            // .collect::<syn::Result<(Vec<_>, Vec<_>)>>()?;
             .collect::<syn::Result<Vec<_>>>()?;
-        let name = name.as_deref().unwrap_or("default_index");
-        indexes.push(quote!(spacetimedb::IndexDesc {
+        let (col_ids, col_tys): (Vec<_>, Vec<_>) = cols.into_iter().unzip();
+        let ident = name.unwrap_or_else(|| {
+            Ident::new(
+                &field_names
+                    .iter()
+                    .map(ToString::to_string)
+                    .collect::<Vec<_>>()
+                    .join("_"),
+                quote!(#(#field_names)*).span(),
+            )
+        });
+        let name = ident.to_string();
+        let index_index = indexes.len();
+        indexes.push(quote!(spacetimedb::table::IndexDesc {
             name: #name,
             ty: spacetimedb::spacetimedb_lib::db::raw_def::IndexType::#ty,
             col_ids: &[#(#col_ids),*],
         }));
-    }
-
-    let (unique_columns, nonunique_columns): (Vec<_>, Vec<_>) =
-        columns.iter().partition(|x| x.attr.contains(ColumnAttribute::UNIQUE));
-
-    let has_unique = !unique_columns.is_empty();
-
-    let mut unique_filter_funcs = Vec::with_capacity(unique_columns.len());
-    let mut unique_update_funcs = Vec::with_capacity(unique_columns.len());
-    let mut unique_delete_funcs = Vec::with_capacity(unique_columns.len());
-    for unique in unique_columns {
-        let column_index = unique.index;
-        let vis = unique.field.vis;
-        let column_type = unique.field.ty;
-        let column_ident = unique.field.ident.unwrap();
-
-        let filter_func_ident = format_ident!("filter_by_{}", column_ident);
-        let update_func_ident = format_ident!("update_by_{}", column_ident);
-        let delete_func_ident = format_ident!("delete_by_{}", column_ident);
-
-        unique_filter_funcs.push(quote! {
-            #vis fn #filter_func_ident(#column_ident: &#column_type) -> Option<Self> {
-                spacetimedb::query::filter_by_unique_field::<Self, #column_type, #column_index>(#column_ident)
-            }
-        });
-
-        unique_update_funcs.push(quote! {
-            #vis fn #update_func_ident(#column_ident: &#column_type, value: Self) -> bool {
-                spacetimedb::query::update_by_field::<Self, #column_type, #column_index>(#column_ident, value)
-            }
-        });
-
-        unique_delete_funcs.push(quote! {
-            #vis fn #delete_func_ident(#column_ident: &#column_type) -> bool {
-                spacetimedb::query::delete_by_unique_field::<Self, #column_type, #column_index>(#column_ident)
+        let index_ty = quote!((#(#col_tys,)*));
+        index_accessors.push(quote! {
+            fn #ident(&self) -> impl spacetimedb::BTreeIndex<Row = #row_type, IndexType = #index_ty>
```

Comment:
I had hoped/thought so, but in practice, no, you have to import the trait to call the methods.

Reply:
Do you think using an `impl Trait` type here is tangibly better in some way than a concrete type? If so, I'm open to revising the proposal, but my intuition is that we should just return a `BTreeIndex<#row_type, #index_ty>`.

Reply:
Hm, I guess that's good - I assumed the proposal meant an autogenerated concrete type, but if it can be a lib-defined concrete type, then that sounds good.

Reply:
The main thing is that we have to expose the implementation detail of the `const COL_IDX: u16` arg, but I think that's fine.

Reply:
oh, that's why I assumed :upside_down_face:

> ```rust
> // Name may vary. Type may be generic over a lifetime bound.
> struct MyTableMyUniqueColumnUniqueIndex {
>     // private fields
> }
> 
> impl TableHandle {
>     // May return a shared reference `&MyTableMyUniqueColumnUniqueIndex`.
>     fn my_unique_column(&self) -> MyTableMyUniqueColumnUniqueIndex;
> }
> ```

Reply:
Ah, this is more than I intended to specify. Returning a concrete library type is fine. I'll make a note to update the proposal.